plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts

```diff
@@ -89,6 +94,15 @@ export class TaskWorker {
     })();
   }
 
+  listen() {
+    this.options.eventEmitter.on('storageTaskBroker:newTask', () => {
+      this.options.taskBroker.claim().then(task => {
+        this.runOneTask(task);
+      });
+    });
```

Comment:
Tbh, I don't think we actually need the event emitter to achieve the non-blocking nature that you're looking for?

If you look at the implementation of the `start` method just above, the `claim()` method will return as soon as a new task is available. It's basically an event listener but one that's driven from the consumer end.

What makes the current implementation blocking is that we wait for `runOneTask` to complete. We do this so that we don't get a runaway effect where the backend gets overloaded due to too many tasks running at the same time. What we could do is to let multiple tasks run concurrently, but add a limit instead. For example, it might pick up and execute tasks up until there are 5 tasks in flight, at which point the implementation waits for one of them to complete before it calls `claim()` again.

Reply:
Hi @Rugvip, thanks for your feedback! @benjdlambert also commented about the eventEmitter in the RFC. I responded to the point in [the comment of #13799](https://github.com/backstage/backstage/issues/13799#issuecomment-1256588769).

Can you talk a little more about the "runaway effect"? My interpretation of "overloading" the system is that the tasks will take up the memory. To me, it's the JS engine's job of allocating memory and managing the timing of the execution with the event loop. One badly placed infinite loop in one task will kill the runtime over 1000 concurrent fetch calls. So, restricting the number of concurrent tasks doesn't prevent overloading the server.

Reply:
It's based on the assumption that there is some amount of CPU or local I/O bound actions that need to run, and that it's better to queue them up instead of running them all at once. To be clear, I'm thinking about an order of 10-100 concurrent task executions.

Either way it doesn't matter much tbh, I think some form of maximum concurrent tasks configuration would make sense. But we can allow that to be set to a very high value, or infinity

Reply:
Hi @Rugvip, as the result of the RFC. I've significantly simplified this PR. I'm going to resolve this comment thread. Thanks!